habitat-lab/habitat/config/default_structured_configs.py

```diff
@@ -1198,23 +1185,23 @@ class DatasetConfig(HabitatBaseConfig):
     type: str = "PointNav-v1"
     split: str = "train"
     scenes_dir: str = "data/scene_datasets"
-    content_scenes: List[str] = field(default_factory=lambda: ["*"])
+    content_scenes: List[str] = ["*"]
     data_path: str = (
         "data/datasets/pointnav/"
         "habitat-test-scenes/v1/{split}/{split}.json.gz"
     )
 
 
-@dataclass
+@attr.s(auto_attribs=True, slots=True)
 class GymConfig(HabitatBaseConfig):
     auto_name: str = ""
     obs_keys: Optional[List[str]] = None
     action_keys: Optional[List[str]] = None
-    achieved_goal_keys: List = field(default_factory=list)
-    desired_goal_keys: List[str] = field(default_factory=list)
+    achieved_goal_keys: List = []
```

Comment:
Apparently, you can just attr.Factory as syntactic sugar: https://www.attrs.org/en/stable/examples.html#defaults

Reply:
It doesn't look like I need to. I looked into the configuration that gets passed to run.py and it looks like all the fields (even lists) are correct.

Reply:
@vincentpierre It works for now, but it's bugprone behvaior. We have rule against it explicitly, but they don't work on attrs:
```
B006: Do not use mutable data structures for argument defaults. They are created during function definition time. All calls to the function reuse this one instance of that data structure, persisting changes between them.
```
^ this also applies to dataclasses and attrs

Reply:
@vincentpierre, it's less likely to be an issue since it would primarily be an issue with rewriting the configs, which we do sparingly, but it could lead to bugprone behavior down the line.

Reply:
Quote from [this comment](https://fb.workplace.com/groups/349226332644221/posts/1224215518478627/?comment_id=1224319071801605)

> Note that when OmegaConf uses the default value to create a ListConfig, it creates a separate container (i.e., adding element to the ListConfig will not add elements to the original Python list it was created from). As a result, it should actually be safe to provide a mutable list as default value (as long as you don't use that class elsewhere...)

Reply:
@vincentpierre Still really yucky if you want to use the dataclasses without Hydra (like for UnitTesting). Also the attrs.slot thing doesn't buy us a whole lot of speed unless we are instantiating the hydra config back into the dataclass.

A better method for speedup is just to use instationate the config when storing it in the class. This converts it into the dataclass which will be faster to access than the DictConfig with or without slots.

Reply:
@vincentpierre You can even just make it pass all the config keys to the class like so: https://hydra.cc/docs/advanced/instantiate_objects/overview/